src/System.Private.CoreLib/src/System/Threading/ManagedThreadId.cs

```diff
@@ -14,108 +14,95 @@
+        //      System::Environment.get_CurrentManagedThreadId
+        //      System::Threading::Lock.Acquire
+        //      System::Runtime::CompilerServices::ClassConstructorRunner::Cctor.GetCctor
+        //      System::Runtime::CompilerServices::ClassConstructorRunner.EnsureClassConstructorRun
+        //      System::Threading::ManagedThreadId.get_Current
+        //      System::Environment.get_CurrentManagedThreadId
+
+        private static ManagedThreadId s_recycledIdsList;
+        private static int s_maxUsedThreadId;
+
+        private int _managedThreadId;
+        private ManagedThreadId _next; // Linked list of recycled ids
+
+        internal const int ManagedThreadIdNone = 0;
+
         public static int Current
         {
             get
             {
                 int currentManagedThreadId = t_currentManagedThreadId;
-                if (currentManagedThreadId == Environment.ManagedThreadIdNone)
+                if (currentManagedThreadId == ManagedThreadIdNone)
                     return MakeForCurrentThread();
                 else
                     return currentManagedThreadId;
             }
         }
 
-        private static volatile ManagedThreadId s_list;
-        private static int s_nextThreadId;
+        private static int MakeForCurrentThread()
+        {
+            ManagedThreadId newManagedThreadId = null;
 
-        private readonly int _managedThreadId; // The ID represented by this object.
-        private volatile int _lock;            // 1 if a thread is trying to recycle this ID; 0 otherwise.
-        private IntPtr _nativeThreadHandle;    // A handle to the thread that currently owns this ID; the thread may be dead.
-        private ManagedThreadId _next;
+            // Try to pop a recycled id from the list
+            for (;;)
+            {
+                var recycledId = Volatile.Read(ref s_recycledIdsList);
+                if (recycledId == null)
+                    break;
 
-        public ManagedThreadId(int id)
-        {
-            _managedThreadId = id;
+                if (Interlocked.CompareExchange(ref s_recycledIdsList, recycledId._next, recycledId) == recycledId)
+                {
+                    GC.ReRegisterForFinalize(recycledId);
+                    newManagedThreadId = recycledId;
+                    break;
+                }
+            }
+
+            if (newManagedThreadId == null)
+            {
+                newManagedThreadId = new ManagedThreadId(Interlocked.Increment(ref s_maxUsedThreadId));
+            }
+
+            t_currentThreadId = newManagedThreadId;
+            t_currentManagedThreadId = newManagedThreadId._managedThreadId;
+            return t_currentManagedThreadId;
         }
 
+        private ManagedThreadId(int managedThreadId)
+        {
+            _managedThreadId = managedThreadId;
+        }
 
-        private static int MakeForCurrentThread()
+        ~ManagedThreadId()
```

Comment:
We should add `if (Environment.HasShutdownStarted) return;` here to avoid returning bogus threadids during shutdown.